### DISCUSSION TITLE
Problems properly activating a plugin on ImHex

### DISCUSSION THREAD

**Discussion by exsilium** [WerWolv/ImHex#1540] [Get Help]
Firstly, big thanks for all the effort that goes into creating and maintaining ImHex, it's an amazing editor and a well structured project! In hopes of trying to extend and contribute to the ecosystem I started to check out the existing plugins in the hopes to create my own. Doing so I ran into some issues. I hope someone can enlighten me.

I thought of writing my own custom provider, looking that there's a 3rd party plugin for Pcaps available, it seemed like a great resource on how to do that. Pcap plugin is based on the Plugin template so it should be quite standard way of creating the `.hexplug`.

Now to the issue at hand. The Pcap plugin works great if I compile it and use the imhex binary that gets produced along with it, but if I take the compiled .hexplug and drop that plug-in to another previously compiled instance of ImHex (same version), I see that the plugin gets initialised successfully (based on log manager output), but the actual Provider does not appear to the UI.

I added a log entry to where the Provider should be added to the list but it feels like this part of the code is never executed, if I just drop the `.hexplug` to another ImHex instance.

```
IMHEX_PLUGIN_SETUP("Pcap Provider", "Dragos", "Plugin for reading pcap files") {
    hex::ContentRegistry::Provider::add<PcapProvider>(true);
    hex::log::info("PcapProvider Added");
}
```

Am i missing something trivial here or advice how to debug this further.

TLDR; I'm having problems bringing plugins across to other ImHex instances, even if they are the same version.

**Comment by WerWolv** [comment]
Hey!

The logs should tell you if anything went wrong when loading your plugin. Could you post it here?
The main things that can go wrong are
- The plugin is in the wrong folder so ImHex doesn't see it
- The plugin depends on some external dlls that can't be found
- The plugin specifies a wrong supported version value causing ImHex to refuse to initialize it (and print an error message to the logs)

**Comment by exsilium** [comment]
Hi @WerWolv , and thank you for your quick reply. Adding some more info here:

<img width="1271" alt="Screenshot 2024-02-10 at 16 31 51" src="https://github.com/WerWolv/ImHex/assets/2134238/dbd15c2b-30dc-4315-9d36-79b8f83ce17d">

From command line I can see also that the plugin is found:

```
117649-Despodent-Pyre build % ./imhex.app/Contents/MacOS/imhex --plugins                           
[16:32:07] [INFO]  [libimhex | Main]           Loading plugin 'pcaps.hexplug'
[16:32:07] [INFO]  [libimhex | Main]           Loading plugin 'fonts.hexpluglib'
[16:32:07] [INFO]  [libimhex | Main]           Loading plugin 'ui.hexpluglib'
[16:32:07] [INFO]  [libimhex | Main]           Loading plugin 'builtin.hexplug'
[16:32:07] [INFO]  [libimhex | Main]           Loading plugin 'disassembler.hexplug'
[16:32:07] [INFO]  [libimhex | Main]           Loading plugin 'decompress.hexplug'
[16:32:07] [INFO]  [libimhex | Main]           Loading plugin 'hashes.hexplug'
[16:32:07] [INFO]  [libimhex | Main]           Loading plugin 'visualizers.hexplug'
[16:32:07] [INFO]  [libimhex | Main]           Loading plugin 'yara_rules.hexplug'
[16:32:07] [INFO]  [libimhex | Main]           Loading plugin 'script_loader.hexplug'
[16:32:07] [INFO]  [libimhex | Main]           Loading plugin 'diffing.hexplug'
Loaded plugins:
- Pcap Provider by Dragos
  Plugin for reading pcap files
- Fonts by Unknown
  
- UI by Unknown
  
- Built-in by WerWolv
  Default ImHex functionality
- Disassembler by WerWolv
  Disassembler support
- Decompressing by WerWolv
  Support for decompressing data
- Hashes by WerWolv
  Hashing algorithms
- Visualizers by WerWolv
  Visualizers for the Pattern Language
- Yara Rules by WerWolv
  Support for matching Yara rules
- Script Loader by WerWolv
  Script Loader plugin
- Diffing by WerWolv
  Support for diffing data
```

One curious thing to add is also that, if I startup ImHex from command line, I can see the additional pcap log entry I added in the terminal, but not captured in the Log Manager. But still, it's not loaded to the effect that the plugin would be listed as one of the Providers:

```
117649-Despodent-Pyre build % ./imhex.app/Contents/MacOS/imhex                                     
[16:38:23] [INFO]  [main | Main]               Welcome to ImHex 1.32.2!
[16:38:23] [INFO]  [main | Main]               Compiled using commit master@09dd25d
[16:38:23] [INFO]  [main | Main]               Running on macOS 23.3.0 Darwin Kernel Version 23.3.0: Wed Dec 20 21:30:44 PST 2023; root:xnu-10002.81.5~7/RELEASE_ARM64_T6000 (arm64)
[16:38:23] [INFO]  [main | Main]               Native scaling set to: 1.0
[16:38:23] [INFO]  [main | Main]               Using 'Apple' GPU
[16:38:23] [INFO]  [main | Init Tasks]         Task 'Setting up environment' finished successfully in 0 ms
[16:38:23] [INFO]  [main | Init Tasks]         Task 'Creating directories' finished successfully in 0 ms
[16:38:23] [INFO]  [main | Init Tasks]         Task 'Loading settings' finished successfully in 0 ms
[16:38:23] [INFO]  [libimhex | Init Tasks]     Loading plugin 'pcaps.hexplug'
[16:38:23] [INFO]  [libimhex | Init Tasks]     Loading plugin 'fonts.hexpluglib'
[16:38:23] [INFO]  [libimhex | Init Tasks]     Loading plugin 'ui.hexpluglib'
[16:38:23] [INFO]  [libimhex | Init Tasks]     Loading plugin 'builtin.hexplug'
[16:38:23] [INFO]  [libimhex | Init Tasks]     Loading plugin 'disassembler.hexplug'
[16:38:23] [INFO]  [libimhex | Init Tasks]     Loading plugin 'decompress.hexplug'
[16:38:23] [INFO]  [libimhex | Init Tasks]     Loading plugin 'hashes.hexplug'
[16:38:23] [INFO]  [libimhex | Init Tasks]     Loading plugin 'visualizers.hexplug'
[16:38:23] [INFO]  [libimhex | Init Tasks]     Loading plugin 'yara_rules.hexplug'
[16:38:23] [INFO]  [libimhex | Init Tasks]     Loading plugin 'script_loader.hexplug'
[16:38:23] [INFO]  [libimhex | Init Tasks]     Loading plugin 'diffing.hexplug'
[16:38:23] [INFO]  [libimhex | Init Tasks]     Library 'fonts.hexpluglib' initialized successfully
[16:38:23] [INFO]  [libimhex | Init Tasks]     Library 'ui.hexpluglib' initialized successfully
[16:38:23] [INFO]  [pcaps]      PcapProvider Added
[16:38:23] [INFO]  [libimhex | Init Tasks]     Plugin 'pcaps.hexplug' initialized successfully
```

In the instance where the plugin actually works, I can also see the `PcapProvider Added` log entry in the graphical Log manager view. 👀 ❔

**Comment by WerWolv** [comment]
The format of the log message looks really weird. Are you potentially linking your plugin to an old version of libimhex?
You should be using this template here: https://github.com/WerWolv/ImHex-Plugin-Template

**Comment by exsilium** [comment]
Could be, thank you for your suggestion, I'll try with the updated template and retry
